src/plugins/ui_feature/find_and_replace.ts

```diff
@@ -138,37 +154,119 @@ export class FindAndReplacePlugin extends UIPlugin {
+        const sheetId = this.getters.getActiveSheetId();
+        const sheetIndex = this.getters.getSheetIds().indexOf(sheetId);
+        this.findMatchesInSheet(sheetId, sheetIndex, matches);
+        break;
+      case "specificRange":
+        const specificRange = this.searchOptions.specificRange;
+        if (specificRange) {
+          let rangeSheet, rangeCells;
+
+          if (specificRange.includes("!")) {
+            [rangeSheet, rangeCells] = specificRange.split("!");
+          } else {
+            rangeSheet = this.getters.getSheetName(this.getters.getActiveSheetId());
+            rangeCells = specificRange;
+          }
+
+          const [startCell, endCell] = rangeCells.split(":");
+
+          const finalEndCell = endCell || startCell;
+          const specificRangeWithEndCell = `${startCell}:${finalEndCell}`;
+
+          const sheets = this.getters.getSheetIds();
+          for (const [sheetIndex, sheet] of sheets.entries()) {
+            const sheetName = this.getters.getSheetName(sheet);
+            const cleanedRangeSheet = rangeSheet.replace(/'/g, "");
+            if (sheetName === cleanedRangeSheet) {
+              this.findMatchesInSheet(sheet, sheetIndex, matches, specificRangeWithEndCell);
+            }
+          }
+        }
+        break;
     }
-    return matches.sort(this.sortByRowThenColumn);
+
+    // Sort the matches by sheet index, row, and column
+    return matches.sort(this.sortBySheetIndexThenRowThenColumn);
   }
 
-  private sortByRowThenColumn(a: Position, b: Position) {
-    if (a.row === b.row) {
-      return a.col - b.col;
+  private findMatchesInSheet(
+    sheetId: string,
+    sheetIndex: number,
+    matches: SearchMatch[],
+    specificRange?: string
+  ) {
+    let top, left, bottom, right;
+
+    if (specificRange) {
+      const rangeData = this.getters.getRangeDataFromXc(sheetId, specificRange);
+      top = rangeData._zone.top;
+      left = rangeData._zone.left;
+      bottom = rangeData._zone.bottom;
+      right = rangeData._zone.right;
+    } else {
+      top = 0;
+      left = 0;
+      bottom = this.getters.getNumberCols(sheetId);
+      right = this.getters.getNumberRows(sheetId);
     }
-    return a.row > b.row ? 1 : -1;
+
+    const cells = this.getters.getCells(sheetId);
+
+    for (const cell of Object.values(cells)) {
+      const { col, row } = this.getters.getCellPosition(cell.id);
+
+      if (col < left || col > right || row < top || row > bottom) {
+        continue;
+      }
+
+      if (
+        cell &&
+        this.currentSearchRegex &&
+        this.currentSearchRegex.test(this.getSearchableString({ sheetId, col, row }))
+      ) {
+        const match: SearchMatch = {
+          sheetId,
+          sheetIndex,
+          col,
+          row,
+          selected: false,
+        };
+        matches.push(match);
+      }
+    }
+  }
+
+  private sortBySheetIndexThenRowThenColumn(a: SearchMatch, b: SearchMatch) {
+    return a.sheetIndex !== b.sheetIndex
```

Comment:
sheetIndex is useless, you can rollback the change to the previous implementation of `sortByRowThenColumn`